contracts/exchange-wrappers/ZeroExExchangeWrapper.sol

```diff
@@ -13,38 +13,104 @@
 
 pragma solidity ^0.8.0;
 
+import { I_ExchangeWrapper } from "../interfaces/I_ExchangeWrapper.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { I_StarkwareContract } from "../interfaces/I_StarkwareContracts.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+/**
+ * @title ZeroExExchangeWrapper
+ * @author dYdX
+ *
+ * Wrapper around 0x API ERC20 token swap.
+ */
+contract ZeroExExchangeWrapper is I_ExchangeWrapper {
+    using SafeERC20 for IERC20;
+
+    // ============ State Variables ============
+
+    IERC20 immutable USDC_ADDRESS;
+
+    uint256 immutable USDC_ASSET_TYPE;
+
+    I_StarkwareContract public immutable STARKWARE_CONTRACT;
+
+    // ============ Constructor ============
+
+    constructor(
+        I_StarkwareContract starkwareContractAddress,
+        IERC20 usdcAddress,
+        uint256 usdcAssetType
+    )
+    {
+        STARKWARE_CONTRACT = starkwareContractAddress;
+        USDC_ADDRESS = usdcAddress;
+        USDC_ASSET_TYPE = usdcAssetType;
+
+        // Set the allowance to the highest possible value.
+        usdcAddress.safeApprove(address(starkwareContractAddress), type(uint256).max);
+    }
 
-import { I_ExchangeWrapper } from "../external/I_ExchangeWrapper.sol";
-import { ERC20 } from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
 
-contract ZeroExExchangeWrapper is I_ExchangeWrapper  {
     // ============ Public Functions ============
 
     /**
-     * Exchange some amount of takerToken for makerToken. Just for tests
+    * Approve an exchange to swap an asset
+    *
+    * @param exchange Address of exchange that will be swapping a token
+    * @param token    Address of token that will be swapped by the exchange
+    */
+    function approveSwap(
+      address exchange,
+      IERC20 token
+    )
+      external
+    {
+      // safeApprove requires unsetting the allowance first.
+      token.safeApprove(exchange, 0);
+      token.safeApprove(exchange, type(uint256).max);
+    }
```

Comment:
This is brittle since we don't know if the 0x api exchange address will always be the same. Unless you're expecting the user to call this function for every single trade, which also isn't good.

Instead see how `approve` is called internally for each trade by this exchange wrapper: https://github.com/dydxprotocol/exchange-wrappers/blob/master/contracts/exchange-wrappers/CurveExchangeWrapper.sol#L70

Reply:
The point of this (and @lucas-dydx) can weigh in is that a user now has the added flexibility if they are calling this on their own to NOT make this call if they know they've already used said exchange. So worst case the contract call costs the same as an internal approve but best case it can cost less